internal/server/server.go

```diff
@@ -1,24 +1,123 @@
 package server
 
 import (
-	"fmt"
+	"crypto/tls"
 	"net/http"
+	"net/http/httputil"
+	"net/url"
+
+	"github.com/kanopy-platform/grafana-auth-proxy/internal/config"
+	"github.com/kanopy-platform/grafana-auth-proxy/internal/jwt"
+	"github.com/kanopy-platform/grafana-auth-proxy/pkg/grafana"
+	log "github.com/sirupsen/logrus"
 )
 
+type GrafanaResponseHeaders struct {
+	User string
+}
+
 type Server struct {
-	router *http.ServeMux
+	router                 *http.ServeMux
+	cookieName             string
+	groups                 config.Groups
+	grafanaProxyUrl        *url.URL
+	grafanaClient          *grafana.Client
+	grafanaResponseHeaders GrafanaResponseHeaders
+	skipTLSVerify          bool
 }
 
-func New() http.Handler {
+type ServerFuncOpt func(*Server) error
+
+func New(opts ...ServerFuncOpt) (http.Handler, error) {
 	s := &Server{router: http.NewServeMux()}
 
+	// load options
+	for _, opt := range opts {
+		err := opt(s)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	s.router.HandleFunc("/", s.handleRoot())
 
-	return s.router
+	return s.router, nil
 }
 
 func (s *Server) handleRoot() http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
-		fmt.Fprintln(w, "hello world")
+		// Extract token from cookie
+		cookie, err := r.Cookie(s.cookieName)
+		if err != nil {
+			logAndError(w, http.StatusUnauthorized, err, "error reading cookie")
+			return
+		}
+
+		// Get claims from token
+		claims, err := jwt.TokenClaims(cookie.Value)
+		if err != nil {
+			logAndError(w, http.StatusUnauthorized, err, "error reading claims from jwt token")
+			return
+		}
+
+		if claims.Subject == "" {
+			logAndError(w, http.StatusUnauthorized, err, "sub claim is empty")
+			return
+		}
```

Comment:
These ifs are uncovered in tests and the logic that follows depends on a number of assumptions about the value of claims.Subject so we should add a test to assert these fail properly.

- No cookie
- invalid jwt
- jwt with an empty Subject claim

Reply:
makes sense